newshomepages/hyperlinks.py

```diff
@@ -52,25 +51,16 @@ def _get_links(context: BrowserContext, data: dict, timeout: int = 180):
 
     # Go to the page
     page.goto(data["url"], timeout=timeout * 1000)
-
-    # Pull the html
-    html = page.content()
-
-    # Parse out all the links
-    soup = BeautifulSoup(html, "html5lib")
-    link_list = soup.find_all("a")
-
-    # Parse out the data we want to keep
-    data_list = []
-    for link in link_list:
-        try:
-            d = dict(text=link.text, url=link["href"])
-        except KeyError:
-            # If no href, skip it
-            continue
-
-        # Add to big list
-        data_list.append(d)
+    # Parse out the data we want to keep in JavaScript
+    link_list = page.evaluate(
+        """
+                              Array.from(
+                                document.getElementsByTagName("a"))
+                                    .map((a) => [a, a.getBoundingClientRect()])
+                                    .map(([a, rect]) => ({"text": a.text, "url": a.href, "x": rect.x, "y": rect.y, "w": rect.width, "h": rect.height })
+                              )"""
+    )
+    data_list = [item for item in link_list if item["url"]]
 
```

Comment:
What do you think of an addition like this, which I hope would raise errors if the JS is failing to find links

```suggestion

# Verify we actually got something back
assert len(data_list) > 0, "No hyperlinks found"

```

Reply:
This will fail in a few cases (like the WSJ, as discussed in that other PR), both with this PR and with the current prod code. That might be okay, just want to flag.

Reply:
Maybe I'm thinking about it wrong, but I think we want it to fail. 

_Errors should never pass silently.
Unless explicitly silenced._

https://en.wikipedia.org/wiki/Zen_of_Python

Reply:
I can see it both ways. `hyperlinks` is accurately reflecting the state of the page (no links, just captchas), but the state of the page is not what we want it to be.

Reply:
Hmm. I'm going to lean loud and vote for the error. I'd rather know it's happening than not. And I don't see the value in saving a bunch of empty lists to archive.org.

Reply:
```
(news-homepags) ➜  news-homepages git:(feature/hyperlink-x-y-w-h) ✗ time python -m newshomepages.hyperlinks wsj
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/jeremybmerrill/code/news-homepages/newshomepages/hyperlinks.py", line 84, in <module>
    cli()
  File "/Users/jeremybmerrill/.pyenv/versions/news-homepags/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jeremybmerrill/.pyenv/versions/news-homepags/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/jeremybmerrill/.pyenv/versions/news-homepags/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jeremybmerrill/.pyenv/versions/news-homepags/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jeremybmerrill/code/news-homepages/newshomepages/hyperlinks.py", line 35, in cli
    link_list = _get_links(context, site, timeout=int(timeout))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jeremybmerrill/.pyenv/versions/news-homepags/lib/python3.11/site-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jeremybmerrill/.pyenv/versions/news-homepags/lib/python3.11/site-packages/retry/api.py", line 73, in retry_decorator
    return __retry_internal(partial(f, *args, **kwargs), exceptions, tries, delay, max_delay, backoff, jitter,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jeremybmerrill/.pyenv/versions/news-homepags/lib/python3.11/site-packages/retry/api.py", line 33, in __retry_internal
    return f()
           ^^^
  File "/Users/jeremybmerrill/code/news-homepages/newshomepages/hyperlinks.py", line 77, in _get_links
    assert len(data_list) > 0, "No hyperlinks found"
           ^^^^^^^^^^^^^^^^^^
AssertionError: No hyperlinks found
```

sounds good that's what it'll look like

Reply:
I like it.